Implement rebootstrap-raft command

babbageclunk opened the pull request:
This will generate an empty raft directory with servers matching the MongoDB replicaset.

Commit:
Implement rebootstrap-raft

Commit:
Tweaks after testing

wallyworld commented on src/rebootstrap-raft/main.go:
```diff
@@ -0,0 +1,290 @@
+	if c.username != "" {
+		info.Database = "admin"
+		info.Username = c.username
+		info.Password = c.password
```
should we warn or error if password is not supplied?

babbageclunk replied:
Yeah, I think this and machineID should be required parameters.

wallyworld reviewed (approved):
Looks ok, assuming it tests live.
Could we add a --dry-run mode that allow it to be run on any server even if the raft dir is still there. It would allow to user to run it anywhere to see what would be done on the problematic server.

babbageclunk commented:
Good idea, I'll add dry-run - it'll just get the replicaset members and show the raft servers it would add.

Commit:
Review + testing changes

Remove username parameter, make it driven by machine-id. This is
needed because the admin password is only available on machine 0 in an
HA cluster of controllers.

Make machine-id and password required.

Add --dry-run.

jameinel commented:
Thanks for doing this. As a side comment, I wonder if it would be easier if
we integrated this as "jujud rebootstrap-raft".
That way the functionality is part of the core tool. And we don't have to
worry about bit rot or finding the tool.

Just a thought. I know mgopurge suffers a bit from not being very
discoverable or integrated.

John
=:->

On Tue, May 7, 2019, 06:09 Christian Muirhead <notifications@github.com>
wrote:

> Merged #1 <https://github.com/juju/rebootstrap-raft/pull/1> into master.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/juju/rebootstrap-raft/pull/1#event-2322749488>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AABRQ7JDX2KYPJDBQEPOK4LPUDQF3ANCNFSM4HLEHUQQ>
> .
>

mitechie commented:
Just a +1 to having it available. I like how we have dump-db behind a feature flag but then available for use. I think following that model would be ideal and set a standard pattern for "hopefully never needed operator tools" going forward.

babbageclunk commented:
Yeah, having it as `jujud rebootstrap-raft` does make sense - it would also mean a lot of the code in here that's ripped from the juju codebase could just be used directly, and we could parse the agent conf rather than having the operator get the password. 
I'll do that once I've got a moment - hopefully after sorting out the other raft issue...
